Basic rspec tests for listnavigation views #2

jzigmund opened the pull request:
Next PR after https://github.com/ManageIQ/manageiq/pull/6956

According to https://github.com/ManageIQ/manageiq/issues/6324

Previously I had issues with these tests, but I was able to fix them so PR is ready to review.

jzigmund commented:
@miq-bot add_label test

jzigmund commented:
@h-kataria as @martinpovolny is on PTO, could you review it ?

martinpovolny commented on spec/controllers/container_group_controller_spec.rb:
```diff
@@ -36,4 +36,25 @@
+    end
+
+    context "render listnav partial" do
+      render_views
```
as with previous PR, we want to make sure that the partila with the listnav renders ok

so there's no point to have it as 2 separate expectations 2 separate contexts

that is unnecessary slow to run and difficult to read

jzigmund replied:
@martinpovolny the point behind separated context is ```render_views```. For getting the HTTP status code, it's not necessary render the views.

How did you find that it's slow  and difficult to read ?
It's not slow, I ran both cases. To readability - this is DSL, so I see almost complete English sentences.

Only valid argument is saving 2 lines if I'll remove context block.

martinpovolny replied:
So you are running the same `get` twice. Once with render views and once w/o render views to do 2 trivial tests.  And you want to do that will all the listnav tests. Why run it twice? What's the benefit?

Do I need to measure how much time it would take to run the extra test w/o render_view for all the listnavs to prove the point?

Readability: It's not difficult to read, but it's less obvious when it's 2 tests compared to having the 2 together. At least for me. Sure enough you can find a differend reviewer with a different idea.

jzigmund replied:
I looked to your PR and I still see there 2 GET requests. Difference is only that you had not separated contexts (if i get you right, you mean this change - martinpovolny/manageiq@037a914)

martinpovolny replied:
I'd be pleased if you can improve the pattern to only do one get per listnav and fix all the tests to follow that pattern.

I merged your changes (and did some fixes) in the previous PR because it gives us at least some coverage in places where we lack it.

martinpovolny commented on app/services/request_referer_service.rb:
```diff
@@ -23,6 +23,8 @@ def self.access_whitelisted?(request, controller_name, action_name)
     'miq_request'              => %w(show:id show_list), # http://localhost:3000/miq_request/show/400r3
     'ems_cluster'              => %w(show:id show_list), # http://localhost:3000/ems_cluster/show/400r2
+    'ems_cloud'                => %w(show:id show_list),
+    'ems_infra'                => %w(show:id show_list),
```
why are you adding this?

jzigmund replied:
Otherwise you cannot directly call the show#id for added resources, the example is failing with error.

martinpovolny replied:
Sure, that is the whitelist for ;-) But do we need to directly access `show#id`?

Or is this added just for the specs to pass (that would be wrong in my opinion).

jzigmund replied:
@martinpovolny I don't know deeper intention of the ENTRY_POINTS_FLAT (only what's written as comment). Is others also wrong? I see that your PR (based on my one), that was merged, is also whitelist new resources. Why did you leave it there if it's wrong in your opinion?

martinpovolny replied:
Because I overlooked it in your changes. Please fix that.

jzigmund commented:
@martinpovolny ah sorry, I thought you're off .. btw. what's wrong with the failed rspecs in Travis log? I don't think it's fault of this PR.

chessbyte commented on app/services/request_referer_service.rb:
```diff
@@ -19,14 +19,11 @@ def self.access_whitelisted?(request, controller_name, action_name)
     'vm'                       => %w(launch_html5_console show:id show_list), # http://localhost:3000/vm/show/400r6
     # 'launch_html5_console' above added due to IE even in version 11 does not set the referrer headir
     # most likely when the new window is opened by JavaScript code
-    'availability_zone'        => %w(show:id show_list), # http://localhost:3000/availability_zone/show/400r1
```
why are we changing the source code if this is spec-only PR?

chessbyte commented on app/services/request_referer_service.rb:
```diff
@@ -19,14 +19,11 @@ def self.access_whitelisted?(request, controller_name, action_name)
     'storage'                  => %w(show:id show_list), # http://localhost:3000//storage/show/400r1
     'ems_container'            => %w(show:id show_list), # http://localhost:3000/ems_container/show/400r1
-    'flavor'                   => %w(show:id show_list),
-    'cloud_tenant'             => %w(show:id show_list), # http://localhost:3000/cloud_tenant/show/400r1
```
why are we changing the source code if this is spec-only PR?

martinpovolny replied:
@chessbyte : we are reverting a previous change for the very reason that you mentioned

I overlooked it previously :-(

chessbyte commented on app/views/layouts/listnav/_ems_container.html.haml:
```diff
@@ -28,6 +28,5 @@
     = miq_accordion_panel(_("Relationships"), false, "ems_container_rel") do
       %ul.nav.nav-pills.nav-stacked
-        - %w(container_project container_route container_service container_replicator container_group ,
-             container container_node container_image_registry container_image persistent_volume container_build).each do |ent|
```
why are we changing the source code if this is spec-only PR?

jzigmund replied:
otherwise the spec example, that's testing a render of the listnav, is failing

chrisarcand commented on spec/factories/container_group.rb:
```diff
@@ -1,5 +1,7 @@
   factory :container_group do
     sequence(:name) { |n| "container_group_#{seq_padded_for_sorting(n)}" }
+    association :ext_management_system, :factory => :ems_kubernetes
+    association :container_project, :factory => :container_project
```
I am wary of this (as everyone should be wary of FactoryGirl associations...) because it doesn't appear that ContainerGroups require these associations. More associations by default == more processing and resources on any specs using this factory that _don't_ depend on these associations. Perhaps you should make them in the spec that they are needed for instead of doing them here?

martinpovolny commented on spec/controllers/container_image_controller_spec.rb:
```diff
@@ -31,6 +31,26 @@
+  describe "#show" do
+    before do
+      EvmSpecHelper.create_guid_miq_server_zone
+      @user = FactoryGirl.create(:user)
```
I donder if we need the `@user` here at all, cannot we just pass that to the `login_as` call?

martinpovolny commented on spec/controllers/container_image_registry_controller_spec.rb:
```diff
@@ -24,6 +24,26 @@
+      @user = FactoryGirl.create(:user)
+      login_as @user
+      @image_registry = FactoryGirl.create(:container_image_registry)
+    end
```
same here

martinpovolny commented on spec/controllers/container_node_controller_spec.rb:
```diff
@@ -23,6 +23,26 @@
+      @user = FactoryGirl.create(:user)
+      login_as @user
+      @node = FactoryGirl.create(:container_node)
+    end
```
and here and or more places....

martinpovolny commented:
Other that the small bit, this looks good!

chrisarcand commented:
@jzigmund Can you please squash these commits before this is merged, also? These are lots of commits to browse history through and are all related to the same 'change' (Adding specs to listnavigation #show views; maybe another commit for the controller changes you made)

jzigmund commented:
updated PR, rebased, merged user creation and ```login_as``` lines to one, merged rspecs to one commit, excluded non-spec changes to separated commits

martinpovolny commented:
Looks good! Please, rebase.
